dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/relationship/hibernate/HibernateRelationshipStore.java

```diff
@@ -75,36 +83,117 @@ public HibernateRelationshipStore( SessionFactory sessionFactory, JdbcTemplate j
 
-        return getList( builder, newJpaParameters( pagingAndSortingCriteriaAdapter, builder )
-            .addPredicate( root -> builder.or(
-                builder.equal( root.join( "from" ).get( "trackedEntityInstance" ), tei ),
-                builder.equal( root.join( "to" ).get( "trackedEntityInstance" ), tei ) ) ) );
+        return getList( relationshipTypedQuery );
     }
 
     @Override
     public List<Relationship> getByProgramInstance( ProgramInstance pi,
         PagingAndSortingCriteriaAdapter pagingAndSortingCriteriaAdapter )
     {
-        CriteriaBuilder builder = getCriteriaBuilder();
+        TypedQuery<Relationship> relationshipTypedQuery = getRelationshipTypedQuery( pi,
+            pagingAndSortingCriteriaAdapter );
 
-        return getList( builder, newJpaParameters( pagingAndSortingCriteriaAdapter, builder )
-            .addPredicate( root -> builder.or(
-                builder.equal( root.join( "from" ).get( "programInstance" ), pi ),
-                builder.equal( root.join( "to" ).get( "programInstance" ), pi ) ) ) );
+        return getList( relationshipTypedQuery );
     }
 
     @Override
     public List<Relationship> getByProgramStageInstance( ProgramStageInstance psi,
         PagingAndSortingCriteriaAdapter pagingAndSortingCriteriaAdapter )
+    {
+        TypedQuery<Relationship> relationshipTypedQuery = getRelationshipTypedQuery( psi,
+            pagingAndSortingCriteriaAdapter );
+
+        return getList( relationshipTypedQuery );
+    }
+
+    private <T extends IdentifiableObject> TypedQuery<Relationship> getRelationshipTypedQuery( T entity,
+        PagingAndSortingCriteriaAdapter pagingAndSortingCriteriaAdapter )
     {
         CriteriaBuilder builder = getCriteriaBuilder();
 
-        return getList( builder, newJpaParameters( pagingAndSortingCriteriaAdapter, builder )
-            .addPredicate( root -> builder.or(
-                builder.equal( root.join( "from" ).get( "programStageInstance" ), psi ),
-                builder.equal( root.join( "to" ).get( "programStageInstance" ), psi ) ) ) );
+        CriteriaQuery<Relationship> relationshipItemCriteriaQuery = builder.createQuery( Relationship.class );
+        Root<Relationship> root = relationshipItemCriteriaQuery.from( Relationship.class );
+
+        setRelationshipItemCriteriaQueryExistsCondition( entity, builder, relationshipItemCriteriaQuery, root );
+
+        return getRelationshipTypedQuery( pagingAndSortingCriteriaAdapter,
+            builder, relationshipItemCriteriaQuery, root );
+    }
+
+    private <T extends IdentifiableObject> void setRelationshipItemCriteriaQueryExistsCondition( T entity,
+        CriteriaBuilder builder,
+        CriteriaQuery<Relationship> relationshipItemCriteriaQuery, Root<Relationship> root )
+    {
+        String relationShipEntityType = getRelationShipEntityType( entity );
+
+        Subquery<RelationshipItem> relationshipItemFromSubQuery = relationshipItemCriteriaQuery
```

Comment:
There is quite a lot of code to read in this method. I would try to keep variable names as short as possible while still being meaningful of course 😋 . We can assume that most of the things are related to relationships as we are in the `HibernateRelationshipsStore`. We also have our IDE for more type information. I would take advantage of these things and shorten var names to for example `fromSubquery` or simply `from` for this one.